agent/processwatcher.go

```diff
@@ -0,0 +1,255 @@
+		buffer.WriteString(fmt.Sprintf("%v\n", proc.String()))
+	}
+	return buffer.String()
+}
+
+type watchedProcess struct {
+	ProcessStarted time.Time
+	Libraries      []library
+	Outdated       bool
+	Pid            int
+}
+
+func (w *watchedProcess) String() string {
+	buffer := bytes.NewBufferString(fmt.Sprintf("PID: %d", w.Pid))
+	if w.Outdated {
+		buffer.WriteString(", is running outdated lib(s)")
+	}
+	for _, lib := range w.Libraries {
+		buffer.WriteString(fmt.Sprintf("\n%v", lib.String()))
+	}
+	return buffer.String()
+}
+
+type library struct {
+	SpectorLib libspector.Library
+	Outdated   bool
+}
+
+func (l *library) String() string {
+	buffer := bytes.NewBufferString("")
+	if l.Outdated {
+		buffer.WriteString("Outdated: yes")
+	} else {
+		buffer.WriteString("Outdated:  no")
+	}
+	buffer.WriteString(fmt.Sprintf(", Path: %v, ", l.SpectorLib.Path()))
+	if pkg, err := l.SpectorLib.Package(); err != nil {
+		buffer.WriteString(fmt.Sprintf("Package error: %v", err))
+	} else {
+		buffer.WriteString(fmt.Sprintf("Package: %v, ", pkg))
+	}
+	if modified, err := l.SpectorLib.Modified(); err != nil {
+		buffer.WriteString(fmt.Sprintf("Modified error: %v", err))
+	} else {
+		buffer.WriteString(fmt.Sprintf("Modified: %v", modified))
+	}
+	return buffer.String()
+}
+
+func (wp *watchedProcess) MarshalJSON() ([]byte, error) {
+	libs := make([]map[string]interface{}, len(wp.Libraries))
+
+	for i, lib := range wp.Libraries {
+		path := lib.SpectorLib.Path()
+
+		modified, err := lib.SpectorLib.Modified()
+		if err != nil {
+			log.Warningf("error retrieving modification date for lib %s, %v", path, err)
+			continue
+		}
+
+		var fullPackage string
+		if pkg, err := lib.SpectorLib.Package(); err != nil {
+			log.Warningf("error retrieving package name for lib %s, %v", path, err)
+		} else {
+			fullPackage = fmt.Sprintf("%s-%s", pkg.Name(), pkg.Version())
+		}
+
+		libs[i] = map[string]interface{}{
+			"path":     path,
+			"modified": modified,
+			"package":  fullPackage,
+			"outdated": lib.Outdated, // in relation to this process
+		}
+	}
+
+	return json.Marshal(map[string]interface{}{
+		"started":   wp.ProcessStarted,
+		"libraries": libs,
+		"outdated":  wp.Outdated,
+		"pid":       wp.Pid,
+	})
+}
+
+func NewProcessWatcher(match string, callback ChangeHandler) Watcher {
+	watcher := &processWatcher{
+		match:     match,
+		OnChange:  callback,
+		UpdatedAt: time.Now(),
+		pollSleep: env.PollSleep,
+	}
+
+	// Don't scan from here, we just end up with two running at once
+	return watcher
+}
+
+func (pw *processWatcher) MarshalJSON() ([]byte, error) {
+	pw.Lock()
+	defer pw.Unlock()
+	return json.Marshal(map[string]interface{}{
+		"match":         pw.Match(),
+		"updated-at":    pw.UpdatedAt,
+		"being-watched": pw.BeingWatched,
+	})
+}
+
+func NewAllProcessWatcher(callback ChangeHandler) Watcher {
+	return NewProcessWatcher("", callback)
+}
+
+func (wt *processWatcher) Start() {
+	wt.Lock()
+	wt.keepPolling = true
+	wt.Unlock()
+	go wt.listen()
```

Comment:
Deferring the unlock until after the go routine is dispatched produced a deadlock most of the time. I'm really not sure about the value of all this coordination.